# **MVP Engenharia de Dados**
**Aluna: Ana Cristina Paula Bomfim**

**MVP Livros mais vendidos na Amazon de 2009 - 2019**


# **Objetivo**

A partir de dados sobre os 50 Livros mais vendidos (Top 50 bestselling) da empresa Amazon no período de 2009-2019, o objetivo deste Pipeline de Dados em questão é responder as seguintes perguntas:

***Pergunta 1***: Qual a quantidade de livros do dataset analisado?

***Pergunta 2***: Quais são os livros que constam neste dataset por Autor? 

***Pergunta 3***: Quais são os gêneros dos livros deste escopo de dados?

***Pergunta 4***: Qual o total de livros para cada gênero neste conjunto de dados?

***Pergunta 5***:Qual o livro com a melhor avaliação de 2019, que custam menos de US$20?**

# **Busca pelo Dados**

O dataset escolhido é uma base de dados gratuita proveniente da plataforma Kaggle, sobre os 50 livros mais vendidos pela empresa Amazon de 2009-2019, e que foram categorizados nos gêneros ficção e não-ficção usando "Goodreads" (Uma plataforma online que funciona como uma rede social e catálogo de livros).

Para mais detalhes sobre o dataset, consulte: https://www.kaggle.com/datasets/sootersaalu/amazon-top-50-bestselling-books-2009-2019

O tipo de licença comun para este conjunto de dados é: Marca de domínio público.

Informação dos campos da Tabela Livros (dataset):

* **Name**: Nome do livro;

* **Autho**r: Autor do livro;

* **UserRating**: Avaliação do usuário;

* **Reviews**: Número de avaliações;

* **Price**: Preço;

* **Year**: Ano em que foi classificado como o mais vendido; 

* **Genre**: Gênero do livro.

# **Coleta**
O Dataset foi baixado da WEB, do site https://www.kaggle.com/code/jaimeandrevelasquez/data-analytics-case-study-amazon-bestsellers/input?select=bestsellers+with+categories.csv), para o disco local no formato de arquivo .CSV, e o PIPELINE DE DADOS em questão coletará este arquivo no referido disco para armazená-lo na nuvem.

**Plataforma**

A plataforma utilizada para o desenvolvimento desse projeto será 
 o Databricks Community.

%md
## Modelagem

Modelo de Dados Flat : Nesse conjunto de dados, os dados são organizados em uma estrutura plana e simples. 
Cada registro ou linha possui uma série de colunas que representam atributos específicos dos dados, como nome, autor, avaliação de usuário, número de avaliações, preço, ano de publicação e gênero, conforme descrito. Não há tabelas adicionais de dimensionamento ou normalização, como ocorre em modelos de dados estrela ou floco de neve.

Descrição do Atributos (Catálogo de Dados) do dataset:

- **Name** : Nome do livro, atributo de texto (STR) representando o título do livro.

- **Author** : Nome do autor do livro, atributo de texto (STR) representando o autor da obra.

- **UserRating** : Avaliação do usuário para o livro, atributo de numérico (FLOAT).

- **Reviews** : Número de avaliações que o livro recebeu, atributo numérico (INT).

- **Price** : Preço do livro, atributo numérico (INT).

- **Year** : Ano em que foi classificado como o mais vendido, atributo numérico (INT). 

- **Genre** : Categoria que descreve o gênero do livro, como ficção ou não-ficção. Atributo de texto (STR)

# **Carga**

Para realizar o processo de ETL na plataforma Databricks Community, foram realizados os seguintes passos.

1- Acessar o Databricks Community Edition.

2- Clicar em "Compute" e criar novo cluster.

3- Importar o dataset para o Databricks através do DBFS.

4- Clicar em "New" e em Table para criar a tabela bestsellers.

5- Clicar em "Create Table With UI".

6- Selecionar o Cluster criado.

7- Clicar em "Preview Table"

8- Configurar os tipos de dados dos campos da tabela.

9- Clicar em "Create Table".

10- Criar um novo notebook e renomear para "MVPEngenhariaDeDados".

In [0]:

# Caminho do arquivo CSV no DBFS
file_path = "dbfs:/FileStore/MVPEngenhariaDeDados/bestsellers_with_categories.csv"

# Carregar o arquivo CSV em um DataFrame Spark
df = spark.read.option("header", "true").csv(file_path)

# Mostrar as primeiras linhas do DataFrame
df.show()

+--------------------+--------------------+-----------+-------+-----+----+-----------+
|                Name|              Author|User Rating|Reviews|Price|Year|      Genre|
+--------------------+--------------------+-----------+-------+-----+----+-----------+
|10-Day Green Smoo...|            JJ Smith|        4.7|  17350|    8|2016|Non Fiction|
|   11/22/63: A Novel|        Stephen King|        4.6|   2052|   22|2011|    Fiction|
|12 Rules for Life...|  Jordan B. Peterson|        4.7|  18979|   15|2018|Non Fiction|
|1984 (Signet Clas...|       George Orwell|        4.7|  21424|    6|2017|    Fiction|
|5,000 Awesome Fac...|National Geograph...|        4.8|   7665|   12|2019|Non Fiction|
|A Dance with Drag...| George R. R. Martin|        4.4|  12643|   11|2011|    Fiction|
|A Game of Thrones...| George R. R. Martin|        4.7|  19735|   30|2014|    Fiction|
|A Gentleman in Mo...|         Amor Towles|        4.7|  19699|   15|2017|    Fiction|
|A Higher Loyalty:...|         James Comey|

In [0]:
%sql
-- Primeiras dez linhas da tabela bestsellers

SELECT * FROM bestsellers LIMIT 10

Name Author User Rating Reviews Price Year Genre 10-Day Green Smoothie Cleanse JJ Smith 4.7 17350 8 2016 Non Fiction 11/22/63: A Novel Stephen King 4.6 2052 22 2011 Fiction 12 Rules for Life: An Antidote to Chaos Jordan B. Peterson 4.7 18979 15 2018 Non Fiction 1984 (Signet Classics) George Orwell 4.7 21424 6 2017 Fiction 5,000 Awesome Facts (About Everything!) (National Geographic Kids) National Geographic Kids 4.8 7665 12 2019 Non Fiction A Dance with Dragons (A Song of Ice and Fire) George R. R. Martin 4.4 12643 11 2011 Fiction A Game of Thrones / A Clash of Kings / A Storm of Swords / A Feast of Crows / A Dance with Dragons George R. R. Martin 4.7 19735 30 2014 Fiction A Gentleman in Moscow: A Novel Amor Towles 4.7 19699 15 2017 Fiction A Higher Loyalty: Truth, Lies, and Leadership James Comey 4.7 5983 3 2018 Non Fiction A Man Called Ove: A Novel Fredrik Backman 4.6 23848 8 2016 Fiction

# Análise

##  **1- Qualidade dos Dados**

Nessa etapa iremos verificar se há necessidade de transformação nos dados da tabela.

Iremos utilizar a linguagem SQL, para análises e transformações.

In [0]:
%sql
-- Verificação Nomes de Livros com nomes Nulos
SELECT * FROM bestsellers
WHERE Name IS NULL

Name Author User Rating Reviews Price Year Genre

In [0]:
%sql
-- Verificação Autores com nomes Nulos
SELECT * FROM bestsellers
WHERE Author IS NULL

Name Author User Rating Reviews Price Year Genre

In [0]:
%sql
-- Verificação Preços com valores negativos.
SELECT * FROM bestsellers
WHERE Price < 0

Name Author User Rating Reviews Price Year Genre

**Com os resultados das análises, verificamos que os dados estão consistentes. Não há necessidade de transformações.**

## **2- Solução do Problema**



**Pergunta 1:** Qual a quantidade de livros do dataset analisado?

In [0]:
%sql
-- Quantidade de livros na tabela
SELECT COUNT(NAME) AS QtdeLivros FROM bestsellers

QtdeLivros 550

O dataset possui 550 livros.

**Pergunta 2:** Quais são os livros que constam neste dataset por Autor?

In [0]:
%sql
SELECT DISTINCT Author, Name FROM bestsellers 
ORDER BY 1 LIMIT 10


Author Name Abraham Verghese Cutting for Stone Adam Gasiewski Milk and Vine: Inspirational Quotes From Classic Vines Adam Mansbach Go the F**k to Sleep Adir Levy What Should Danny Do? (The Power to Choose Series) Admiral William H. McRaven Make Your Bed: Little Things That Can Change Your Life...And Maybe the World Adult Coloring Book Designs Adult Coloring Book Designs: Stress Relief Coloring Book: Garden Designs, Mandalas, Animals, and Paisley Patterns Alan Moore Watchmen Alex Michaelides The Silent Patient Alice Schertle Little Blue Truck Allie Brosh Hyperbole and a Half: Unfortunate Situations, Flawed Coping Mechanisms, Mayhem, and Other Things That Happened

A quantidade de livros apresentada no resultado foi limitada a 10, devido a dificuldade de formatação do resultado após a exportação do projeto.

É possível visualizar os livros por autor. Os livros foram listados pelo nome do autor em ordem alfabética crescente.

**Pergunta 3:** Quais são os gêneros dos livros deste escopo de dados?

In [0]:
%sql
SELECT DISTINCT GENRE FROM bestsellers

GENRE Fiction Non Fiction

Os gêneros do livros listados no dataset são Ficção e Não Ficção.

***Pergunta 4***: Qual o total de livros para cada gênero neste conjunto de dados?

In [0]:
%sql
SELECT Genre, COUNT (GENRE) AS QtdePorGenero FROM bestsellers
group by Genre

Genre QtdePorGenero Fiction 240 Non Fiction 310

Dos 550 livros listados no dataset, 240 são do gênero Ficção e 310 do gênero Não Ficção.

**Pergunta 5:** Quais os livros com a melhor avaliação de 2019, que custam menos de US$20?

In [0]:
%sql
-- Livros listados abaixo.
SELECT NAME, year, `User Rating`, Price FROM bestsellers
WHERE `User Rating`= (SELECT MAX(`User Rating`)FROM bestsellers where year = 2019)
AND Price <= 20
AND Year = 2019
ORDER BY Price ASC

NAME year User Rating Price Dog Man: Brawl of the Wild: From the Creator of Captain Underpants (Dog Man #6) 2019 4.9 4 Brown Bear, Brown Bear, What Do You See? 2019 4.9 5 The Very Hungry Caterpillar 2019 4.9 5 Strange Planet (Strange Planet Series) 2019 4.9 6 Dog Man: Fetch-22: From the Creator of Captain Underpants (Dog Man #8) 2019 4.9 8 Dog Man: For Whom the Ball Rolls: From the Creator of Captain Underpants (Dog Man #7) 2019 4.9 8 Oh, the Places You'll Go! 2019 4.9 8 Wrecking Ball (Diary of a Wimpy Kid Book 14) 2019 4.9 8 The Wonderful Things You Will Be 2019 4.9 10 Unfreedom of the Press 2019 4.9 11 Harry Potter and the Goblet of Fire: The Illustrated Edition (Harry Potter, Book 4) (4) 2019 4.9 18

# **AutoAvaliação**

De modo geral, as perguntas formuladas no inicio do trabalho foram bem respondidas. 

Acredito ter atingido todos os objetivos propostos no trabalho.
Outro ponto fundamental foi o aprendizado na ferramenta Databricks, que está sendo cada vez mais utilizada e solicitada no mercado de trabalho. 

Minha maior dificuldade foi aprender a ferramenta, pois não tinha conhecimento algum dela, e por ser uma versão não paga, o cluster dura apenas 2hs, o que atrasa a finalização do projeto.